CSharp/JobPrep/Program.cs

```diff
@@ -0,0 +1,333 @@
+
+        private static void Log(string msg, string vm = "")
+        {
+#if VERBOSE
+            Console.WriteLine(DateTime.Now.ToString("HH:mm:ss.ffff ") + vm + " " + msg);
+#endif
+        }
+
+        private static bool WaitAllJobPrepTask(IBatchClient client, string pool, string workitem, string job, Func<int, int, bool> stop, Func<int, int, bool> succeed, int timeout = 60*1000, int freq = 2000, int parallelism = 5)
+        {
+            using (IPoolManager pm = client.OpenPoolManager())
+            {
+                using (IWorkItemManager wm = client.OpenWorkItemManager())
+                {
+
+                    CancellationTokenSource cts = new CancellationTokenSource();
+                    CancellationToken ct = cts.Token;
+
+                    BlockingCollection<string> vms = new BlockingCollection<string>();
+                    List<Task> tasks = new List<Task>();
+                    int nb_succeeded = 0;
+                    int nb_failed = 0;
+
+                    // find all VM in the Pool and put them in a queue
+                    Log("add vms");
+                    tasks.Add(Task.Run(async () =>
+                    {
+                        Log("start adding vms");
+                        // list all vms and enqueue
+                        ICloudPool p = pm.GetPool(PoolName);
+                        var en = p.ListVMs(new ODATADetailLevel(selectClause: "name")).GetAsyncEnumerator();
+                        while(await en.MoveNextAsync())
+                        {
+                            Log("add vm", en.Current.Name);
+                            vms.Add(en.Current.Name);
+                            if (ct.IsCancellationRequested)
+                                break;
+                        }
+                    }));
+
+                    // launch N threads to get job prep task status of the nodes in the queue
+                    for (int i = 0; i < parallelism; i++)
+                        tasks.Add(Task.Run(() =>
+                        {
+                            while (!ct.IsCancellationRequested)
+                            {
+                                string vm;
+                                try {
+                                    vm = vms.Take(ct);
+                                } catch // if token cancelled, abort the thread
+                                {
+                                    break;
+                                }
+                                Log("start polling vm", vm);
+                                try
+                                {
+                                    var state = pm.GetJobPreparationTaskStatus(pool, vm, workitem, job);
+                                    Log("got vm state", vm);
+                                    if (state.State == JobPreparationTaskState.Completed) // job prep task completed on the thread
+                                    {
+                                        if (state.ExitCode == 0)
```

Comment:
It looks like you could reorganize the code a bit here to use the PollJobPrepTaskAsync() method you wrote above?
Maybe I missed it, but it doesn't look like you actually invoke that method anywhere.